runtime/src/vote_account.rs

```diff
@@ -83,37 +83,43 @@ impl VoteAccounts {
         self.staked_nodes.read().unwrap().clone()
     }
 
-    pub fn iter(&self) -> impl Iterator<Item = (&Pubkey, &(u64, VoteAccount))> {
+    pub fn get(&self, pubkey: &Pubkey) -> Option<&(/*stake:*/ u64, VoteAccount)> {
+        self.vote_accounts.get(pubkey)
+    }
+
+    pub(crate) fn iter(&self) -> impl Iterator<Item = (&Pubkey, &(u64, VoteAccount))> {
         self.vote_accounts.iter()
     }
 
-    pub fn insert(&mut self, pubkey: Pubkey, (stake, vote_account): (u64, VoteAccount)) {
+    pub(crate) fn insert(&mut self, pubkey: Pubkey, (stake, vote_account): (u64, VoteAccount)) {
         self.add_node_stake(stake, &vote_account);
-        if let Some((stake, vote_account)) =
-            self.vote_accounts.insert(pubkey, (stake, vote_account))
-        {
+        let vote_accounts = Arc::make_mut(&mut self.vote_accounts);
+        if let Some((stake, vote_account)) = vote_accounts.insert(pubkey, (stake, vote_account)) {
             self.sub_node_stake(stake, &vote_account);
         }
     }
 
-    pub fn remove(&mut self, pubkey: &Pubkey) -> Option<(u64, VoteAccount)> {
-        let value = self.vote_accounts.remove(pubkey);
-        if let Some((stake, ref vote_account)) = value {
+    pub(crate) fn remove(&mut self, pubkey: &Pubkey) -> Option<(u64, VoteAccount)> {
+        let vote_accounts = Arc::make_mut(&mut self.vote_accounts);
+        let entry = vote_accounts.remove(pubkey);
+        if let Some((stake, ref vote_account)) = entry {
             self.sub_node_stake(stake, vote_account);
         }
-        value
+        entry
     }
 
-    pub fn add_stake(&mut self, pubkey: &Pubkey, delta: u64) {
-        if let Some((stake, vote_account)) = self.vote_accounts.get_mut(pubkey) {
+    pub(crate) fn add_stake(&mut self, pubkey: &Pubkey, delta: u64) {
+        let vote_accounts = Arc::make_mut(&mut self.vote_accounts);
+        if let Some((stake, vote_account)) = vote_accounts.get_mut(pubkey) {
             *stake += delta;
             let vote_account = vote_account.clone();
             self.add_node_stake(delta, &vote_account);
         }
     }
 
-    pub fn sub_stake(&mut self, pubkey: &Pubkey, delta: u64) {
-        if let Some((stake, vote_account)) = self.vote_accounts.get_mut(pubkey) {
+    pub(crate) fn sub_stake(&mut self, pubkey: &Pubkey, delta: u64) {
+        let vote_accounts = Arc::make_mut(&mut self.vote_accounts);
+        if let Some((stake, vote_account)) = vote_accounts.get_mut(pubkey) {
```

Comment:
Nice! I think this works

Quick sanity check, this should be safe even if we have interleaving updates between Replay and BankingStage.

For example, imagine we have:

```
     /--------5 (Banking Stage)
0
   \-----------4 (ReplayStage)
```

So the sequence of events:
1)  We start a leader block for our own slot `5` , inheriting the shared `stakes` object from the slot `0`: https://github.com/solana-labs/solana/blob/84db04ce6c9bce7e1bb556ec44ec26e0461af5dc/runtime/src/bank.rs#L1344.

2) At some later point in time Replay concurrently creates and starts replaying block `4`, also inheriting the shared `stakes` object from the slot `0`. 

The situation we want to avoid is either of the slot-specific updates to `stakes` via `update_cached_accounts()`: https://github.com/solana-labs/solana/blob/84db04ce6c9bce7e1bb556ec44ec26e0461af5dc/runtime/src/bank.rs#L5152 from being visible to the other child. 

Now we know this is safe because if we do an update in slot `5` before or after `4` is created, we clone the vote accounts for `5`, so these updates are hidden from `4`.

Reply:
Yes, it will be safe because as soon as any mutations are about to be made (and there is still another Arc reference to the same hash-map), a clone is made and so the mutations are not leaked to other Arc references.
The added `test_staked_nodes_cow` test is testing one such scenario.